src/iosource/Manager.cc

```diff
@@ -59,124 +109,411 @@ IOSource* Manager::FindSoonest(double* ts)
+	memset(&event, 0, sizeof(epoll_event));
+	event.events = EPOLLIN;
+	event.data.fd = fd;
 
-		src->Clear();
-		src->src->GetFds(&src->fd_read, &src->fd_write, &src->fd_except);
-		src->SetFds(&fd_read, &fd_write, &fd_except, &maxx);
+	int ret = epoll_ctl(event_queue, EPOLL_CTL_ADD, fd, &event);
+	if ( ret != -1 )
+		{
+		DBG_LOG(DBG_MAINLOOP, "Registered fd %d from %s", fd, src->Tag());
+		events.push_back({});
+		fd_map[fd] = src;
+		Wakeup("RegisterFd");
+		}
+	else
+		{
+		DBG_LOG(DBG_MAINLOOP, "Failed to register fd %d from %s: %s", fd, src->Tag(), strerror(errno));
 		}
+	}
 
-	// We can't block indefinitely even when all sources are dry:
-	// we're doing some IOSource-independent stuff in the main loop,
-	// so we need to return from time to time. (Instead of no time-out
-	// at all, we use a very small one. This lets FreeBSD trigger a
-	// BPF buffer switch on the next read when the hold buffer is empty
-	// while the store buffer isn't filled yet.
+void Manager::UnregisterFd(int fd)
+	{
+	if ( fd_map.find(fd) != fd_map.end() )
+		{
+		int ret = epoll_ctl(event_queue, EPOLL_CTL_DEL, fd, NULL);
+		if ( ret != -1 )
+			DBG_LOG(DBG_MAINLOOP, "Unregistered fd %d", fd);
 
-	struct timeval timeout;
+		events.pop_back();
+		fd_map.erase(fd);
 
-	if ( all_idle )
+		Wakeup("UnregisterFd");
+		}
+	}
+
+bool Manager::Poll(std::set<IOSource*>& ready, double timeout, IOSource* timeout_src)
+	{
+	bool timer_expired = false;
+
+	// TODO: double comparisons are fraught with danger
+	if ( timeout != 0 )
 		{
-		// Interesting: when all sources are dry, simply sleeping a
-		// bit *without* watching for any fd becoming ready may
-		// decrease CPU load. I guess that's because it allows
-		// the kernel's packet buffers to fill. - Robin
-		timeout.tv_sec = 0;
-		timeout.tv_usec = 20; // SELECT_TIMEOUT;
-		select(0, 0, 0, 0, &timeout);
+		struct itimerspec new_timeout;
+		memset(&new_timeout, 0, sizeof(new_timeout));
+		ConvertTimeout(timeout, new_timeout.it_value);
+		timerfd_settime(timerfd, 0, &new_timeout, NULL);
 		}
 
-	if ( ! maxx )
-		// No selectable fd at all.
-		goto finished;
+	// Because of the way timerfd works, you can't just set it to a zero
+	// timeout. That deactivates the timer. That means if the timeout
+	// passed in was zero, we need to pass that zero down to poll().
+	// Otherwise, set it to -1 and let timerfd do its thing.
+	int ret = epoll_wait(event_queue, events.data(), events.size(), timeout == 0 ? 0 : -1);
+	if ( ret == -1 )
+		{
+		// TODO: something better here than just a perror
+		if ( errno != EINTR )
+			perror("poll error");
+		}
+	else if ( ret == 0 )
+		{
+		if ( timeout_src )
+			ready.insert(timeout_src);
+		else
+			timer_expired = true;
+		}
+	else
+		{
+		DBG_LOG(DBG_MAINLOOP, "epoll says %d things are ready", ret);
+		for ( int i = 0; i < ret; i++ )
+			{
+			DBG_LOG(DBG_MAINLOOP, "%d is ready, status is %d", events[i].data.fd, events[i].events);
+			// This theoretically shouldn't be possible, but best to check
```

Comment:
Not sure if I follow correctly, I would have expected the opposite: As long as there's one source ready that's not the timeout_src, we'd want ignore the timeout_src even if it's ready (because that timeout_src may not expect to called before the timeout it specified)

Reply:
Removed all of this with the changes to TimerMgr.